### ISSUE TITLE
--embed-thumbnail for Opus files

### ISSUE THREAD

**Issue by pylorak** [ytdl-org/youtube-dl#22338]
## Checklist

- [x] I'm reporting a feature request
- [x] I've verified that I'm running youtube-dl version **2019.09.01**
- [x] I've searched the bugtracker for similar feature requests including closed ones

## Description

Please add support for embedding cover art in opus files.

**Comment by nullvideo** [comment]
You could externally, use `opusenc` to embed thumbnails into a OPUS files.
I’d try to implement that in YouTubeDL.

**Comment by pylorak** [comment]
For the same reason that youtube-dl already supports embedding cover art into MP4, supporting it for Opus would be a lot more convenient for the user than having to post-process the file with a separate application like opusenc. Also, it would avoid having to re-encode the file just for the sake of adding a cover art.

**Comment by nullvideo** [comment]
I was quite wrong, `opusenc` cannot post-process though, it requires a WAVE/PCM as input to encode to OPUS. This is results in lossy-to-lossy compression (quantization on already quantized signal) which degrades the quality even more.

> OPUS doesn't define it's own bit-stream format, it's usually contained inside a **OGG** container, `opusenc` follows this way.

---

MP4 (*MPEG-4 Part-II*) supports OPUS as a audio-codec. M4A is MP4 (Apple used this name to distinguish between Video and Audio files, besides that the structure is exactly the same).

Boils down to, `AtomicParsley` could be used to merge MP4 covers.

**Comment by milahu** [comment]
.... or keep the ogg container, and embed the cover art image as base64 comment like

```
METADATA_BLOCK_PICTURE=base64_string
```

https://superuser.com/questions/169151/embed-album-art-in-ogg-through-command-line-in-linux

**Comment by tryamid** [comment]
> 
> 
> .... or keep the ogg container
> and embed the cover art image as base64 comment like
> 
> ```
> METADATA_BLOCK_PICTURE=base64_string
> ```
> 
> https://superuser.com/questions/169151/embed-album-art-in-ogg-through-command-line-in-linux

Appending a *cover art* this way has its limitations (eg. argument list can become too long that the current shell might now allow it). Also, this isn’t automation in anyway, suppose you downloaded a bunch of files, you have to iterate through all the files and download it’s corresponding thumbnail from the site where it was downloaded so at the end *things become very messy*.

I recently focused a bit on this problem and found out that [`mutagen`][1] as optional PyPI dependency to embed thumbnails into OGGOpus files would be great.

[1]: https://pypi.org/project/mutagen/

**Comment by pylorak** [comment]
Thank you! I'll test and report back as soon as a Windows build of the new version published.

**Comment by mani445** [comment]
Thanks @tripulse asking for this feature.

**Comment by milahu** [comment]
> I’ve finished the work of implementation

thanks for implementing my suggestion ; ) and thanks for finding a good library for the job

```py

                aufile['metadata_block_picture'] = \
                    base64.b64encode(covart.write()).decode('ascii')
```

> Also, this isn’t automation in anyway, suppose you downloaded a bunch of files, you have to iterate through all the files and download it’s corresponding thumbnail from the site where it was downloaded so at the end *things become very messy.*

i still have thousands of `*.opus` files who need postprocessing (and ♫ im not the only one ♫). if you have some spare time, it would be awesome if you could add a `opus-embed-thumbs.py` script that `glob`s all opus files in a folder and ....

the album cover / cover art / video thumbnail is stored as image file with the same basename (`youtube-dl --embed-thumbnail` keeps the image file when embedding fails)

in case there is no cover art, i still have the video ID stored in the filename (or in the json file produced by `youtube-dl --write-info-json`), so the cover image is just one http request away

**Comment by mani445** [comment]
> > I’ve finished the work of implementation
When work

**Comment by tripulse** [comment]
Well since the PR isn't being merged any sooner, you can do that as a workaround.

**Comment by rautamiekka** [comment]
> Well since the PR isn't being merged any sooner, you can do that as a workaround.

The PR doesn't change the code that much, so you can simple download the repo, make the changes accordingly, and then just execute `youtube_dl/__main__.py` with your flags/params.

**Comment by tripulse** [comment]
For now [ogg-flac-opus.txt](https://github.com/ytdl-org/youtube-dl/files/6390383/ogg-flac-opus.txt) patch can be applied.

**Comment by shmerl** [comment]
> I was quite wrong, `opusenc` cannot post-process though, it requires a _WAV_/_PCM_ as input to encode to OPUS. This is results in lossy-to-lossy compression (quantization on already quantized signal) which degrades the quality even more.

Opened an [issue about it](https://github.com/xiph/opus/issues/244).

**Comment by shmerl** [comment]
See also this: https://github.com/Sound-Linux-More/opustags

**Comment by Natetronn** [comment]
I recently discovered the embed-thumbnail flag and would love to see it function for opus as well.

**Comment by 89z** [comment]
Since its not been mentioned yet, I will. Even if this is implemented, Windows does not support this. Even outside the context of YouTube-DL, Windows does not support thumbnail for Opus files.

**Comment by mYnDstrEAm** [comment]
> Windows does not support thumbnail for Opus files.

What do you mean by this? That it does not support adding thumbnails or playing opus files that have thumbnails?

Either way (or if it's both): wouldn't that be an issue with some Microsoft media player, not Windows? Ffmpeg can also be installed on Windows and it would be used for thumbnail-embedding only if available (there could be a message in the console prompting the user to install it if one would like to have thumbnails).

**Comment by 89z** [comment]
> What do you mean by this? That it does not support adding thumbnails or playing opus files that have thumbnails?

Windows Explorer does not recognized Opus thumbnails, so Opus files will use a generic icon, regardless of any thumbnail.

> Either way (or if it's both): wouldn't that be an issue with some Microsoft media player, not Windows?

No, it would be an issue with Windows Explorer

> Ffmpeg can also be installed on Windows and it would be used for thumbnail-embedding only if available (there could be a message in the console prompting the user to install it if one would like to have thumbnails).

It doesn't matter what program is used, as Windows cannot read the Opus thumbnails, even if they exist.

**Comment by Pauan** [comment]
> It doesn't matter what program is used, as Windows cannot read the Opus thumbnails, even if they exist.

Even so, it would still be useful to have on Windows, for archival purposes. The file might be downloaded on Windows, but then shared with Mac/Linux machines.

**Comment by 89z** [comment]
@Pauan yes good point. I'm not against the feature, I just want it to be very clear beforehand what people should expect out of this.

**Comment by dirkf** [comment]
Is the Windows Explorer issue that it doesn't know about the OGG container and/or Opus stream format, and hence can't extract any thumbnail? If so, the solution is (used to be, anyway) a shell extension that supports the necessary shell icon COM interfaces.

**Comment by 89z** [comment]
> Is the Windows Explorer issue that it doesn't know about the OGG container and/or Opus stream format, and hence can't extract any thumbnail? If so, the solution is (used to be, anyway) a shell extension that supports the necessary shell icon COM interfaces.

Looks like Windows in general has no notion of Opus files. It does know about WebP image files, which is interesting. I just tried this file:

~~~
youtube-dl -x 0bAeJ5fJ1M0
~~~

and Windows Media Player would not play it (I normally use MPC-HC), and should I embed a thumbnail, Windows Explorer wont recognize it. Personally I think any embedding should favor formats with cross platform support (`.mp4`), however I can agree that its also attractive to be able to use the native audio format (`.opus`).

**Comment by rautamiekka** [comment]
Does the [Web Media Extensions](https://www.microsoft.com/en-us/p/web-media-extensions/9n5tdp8vcmhs) off Store help at all ? My Win10 system is modified to remove most things (I use VLC), so I can't test it.

**Comment by 89z** [comment]
When I click "Get", it opens the Microsoft Store, and just loads forever. So I don't know. Why Microsoft thinks a store is needed, I also don't know. A simple link to `.exe` or `.zip` would work just fine.

**Comment by rautamiekka** [comment]
Same as Steam: centralization, auto-updates. If anything, the Store is a smart move, but like everything M$ does, they never do anything right by themselves.

I think you need an M$ acc, and of course you need the Store app to work on your Win10 or 11.

**Comment by 89z** [comment]
I think I installed it, but I cant tell. The interface is awful:

> ![Untitled](https://user-images.githubusercontent.com/73562167/167272373-d08c0a22-494f-4703-8f48-d618af637baa.png)

That is scrolled all the way up. The menu covers up the status of the install.

**Comment by 89z** [comment]
I see now. In "Downloads and updates" it says "Error", "Turn on Windows Update". I am not going to do that, so I guess I will have to find another option.

**Comment by avidseeker** [comment]
This seems to be fixed in v2024.07.01

**Comment by dirkf** [comment]
Are you perhaps running _yt-dlp_? No yt-dl version tagged 2024-07-01 exists.
